Backend.md

```diff
@@ -0,0 +1,167 @@
+
+Shader view "shapes", as in the amount of arguments, and the amount of inputs per argument, are supplied when creating a `handle::pipeline_state`. The actual argument values are supplied in commands, like `cmd::draw`.
+
+## Threading
+
+With one exception, the PRB backend is entirely free-threaded and internally synchronized. The synchronization is minimal, and parallel recording of command lists is encouraged, which takes place on thread-local components. The exception is the `Backend::submit` method, which must only be called on one thread at a time.
+
+## Resource States
+
+PRB exposes a simplified version of resource states, especially when compared with Vulkan. Additionally, resource transitions only ever specify the after-state, the before-state is internally tracked (including thread- and record-order safety).
+
+When transitioning towards `shader_resource` or `unordered_access`, the shader stage(s) which will use the resource next must also be specified.
+
+## Main Loop
+
+In the steady state, there is very little interaction with the backend apart from `command_list` recording and submission. A prototypical PRB main loop looks like this:
+
+```C++
+while (window_open)
+{
+    if (window_resized)
+        backend.onResize({w, h});
+
+    if (backend.clearPendingResize())
+    { // the backbuffer has resized, re-create size dependant resources
+        // think of this event as entirely independent from a window resize, do all your resize logic here
+        // use backend.getBackbufferSize()
+    }
+
+    // ... write and record main command lists ...
+
+    auto const backbuffer = backend.acquireBackbuffer(); // as late as possible
+
+    if (!backbuffer.is_valid()) 
+    { // backbuffer acquire has failed, discard this frame
+        backend.discard(cmdlists);
+        continue;
+    }
+
+    // ... write and record final present command list ...
+    // ... transition backbuffer to resource_state::present ...
+
+    backend.submit(cmdlists);
+    backend.present();
+}
+```
+
+## Vulkan and D3D12 Specifics
+
+### Shader Compilation
+
+All shaders passed to PRB must be in binary format: DXIL for D3D12, SPIR-V for Vulkan. Shaders are to be compiled from HLSL, using [DirectXShaderCompiler](https://github.com/microsoft/DirectXShaderCompiler/releases) (find Linux binaries [here](https://github.com/project-arcana/arcana-samples/tree/develop/res/pr/liveness_sample/shader/dxc_bin/linux) or use [docker](https://hub.docker.com/r/gwihlidal/dxc/)).
+
+When compiling HLSL to SPIR-V for Vulkan, use these flags: `-spirv -fspv-target-env=vulkan1.1 -fvk-b-shift 0 all -fvk-t-shift 1000 all -fvk-u-shift 2000 all -fvk-s-shift 3000 all`. If it is a vertex, geometry or domain shader, the `-fvk-invert-y` must also be added.
+
+For Vulkan, the shader `main` function must be named as such: Vertex - `main_vs`, hull - `main_hs`, domain - `main_ds`, geometry - `main_gs`, pixel - `main_ps`, compute - `main_cs`.
```

Comment:
This doesn't sound ergonomic. Maybe we find a better solution mid-term.

Reply:
The short term solution currently is a [bash script](https://github.com/project-arcana/arcana-samples/blob/kw/develop/res/pr/liveness_sample/shader/dxc_bin/linux/compile_single_shader.sh) which takes just three arguments: domain (vs/ps/cs/...), input, and output file. In the samples, this script is automatically called for each line in [this list](https://github.com/project-arcana/arcana-samples/blob/kw/develop/res/pr/liveness_sample/shader/shader_list.txt) (an analogous windows script exists as well).

As we'll require programmatic shader compilation for the frontend / in the future, we likely want to wrap DXC in our own project, which provides an executable and a library component.

For entry point naming, we could make it an explicit argument to the prb API or possibly additionally reflect it from SPIR-V.